backend/src/apiserver/config/config.go

```diff
@@ -0,0 +1,197 @@
+				File:        cfg.File,
+				Description: cfg.Description,
+			})
+		}
+		pipelineConfig.LoadSamplesOnRestart = false
+	}
+
+	if !pipelineConfig.LoadSamplesOnRestart && haveSamplesLoaded {
+		glog.Infof("Samples already loaded in the past. Skip loading.")
+		return nil
+	}
+
+	for _, cfg := range pipelineConfig.Pipelines {
+		reader, configErr := os.Open(cfg.File)
+		if configErr != nil {
+			return fmt.Errorf("failed to load sample %s. Error: %v", cfg.Name, configErr)
+		}
+		pipelineFile, configErr := server.ReadPipelineFile(cfg.File, reader, common.MaxFileLength)
+		if configErr != nil {
+			return fmt.Errorf("failed to load sample %s. Error: %v", cfg.Name, configErr)
+		}
+
+		// Create pipeline if it does not already exist
+		p, fetchErr := resourceManager.GetPipelineByNameAndNamespace(cfg.Name, "")
+		if fetchErr != nil {
+			if util.IsUserErrorCodeMatch(fetchErr, codes.NotFound) {
+				p, configErr = resourceManager.CreatePipeline(&model.Pipeline{
+					Name:        cfg.Name,
+					Description: cfg.Description,
+				})
+				if configErr != nil {
+					// Log the error but not fail. The API Server pod can restart and it could potentially cause
+					// name collision. In the future, we might consider loading samples during deployment, instead
+					// of when API server starts.
+					glog.Warningf(fmt.Sprintf(
+						"Failed to create pipeline for %s. Error: %v", cfg.Name, configErr))
+					continue
+				}
+			} else {
+				return fmt.Errorf(
+					"Failed to handle load sample for Pipeline: %s. Error: %v", cfg.Name, fetchErr)
+			}
+		}
+
+		// Use Pipeline Version Name/Description if provided
+		// Otherwise fallback to owning Pipeline's Name/Description
+		pvDescription := cfg.Description
+		if cfg.VersionDescription != "" {
+			pvDescription = cfg.VersionDescription
+		}
+		pvName := cfg.Name
+		if cfg.VersionName != "" {
+			pvName = cfg.VersionName
+		}
+
+		// If the Pipeline Version exists, do nothing
+		// Otherwise upload new Pipeline Version for
+		// this pipeline.
+		_, fetchErr = resourceManager.GetPipelineVersionByName(pvName)
+		if fetchErr != nil {
+			if util.IsUserErrorCodeMatch(fetchErr, codes.NotFound) {
+				_, configErr = resourceManager.CreatePipelineVersion(
+					&model.PipelineVersion{
+						Name:         pvName,
+						Description:  pvDescription,
+						PipelineId:   p.UUID,
+						PipelineSpec: string(pipelineFile),
+					},
+				)
+				if configErr != nil {
+					// Log the error but not fail. The API Server pod can restart and it could potentially cause name collision.
+					// In the future, we might consider loading samples during deployment, instead of when API server starts.
+					glog.Warningf(fmt.Sprintf("Failed to create pipeline for %s. Error: %v", pvName, configErr))
+
+					continue
+				}
+				// Since the default sorting is by create time,
+				// Sleep one second makes sure the samples are
+				// showing up in the same order as they are added.
+				time.Sleep(1 * time.Second)
```

Comment:
Optional: I think you only need to do this if there are multiple pipeline versions in a pipeline. So perhaps you can skip the sleep if it's first pipeline version for that pipeline ID.

It'd make the startup a bit faster.

Reply:
agreed, updated!